crates/kornia-vlm/src/smolvlm/mod.rs

```diff
@@ -0,0 +1,184 @@
+
+        let (model, tokenizer) = Self::load_model(dtype, &device)?;
+        let image_token = tokenizer.encode("<image>", false)?;
+        let image_token_tensor = Tensor::from_slice(image_token.get_ids(), &[1], &device)?;
+
+        Ok(Self {
+            model,
+            tokenizer,
+            image_token_tensor,
+            config,
+            logits_processor: LogitsProcessor::new(
+                config.seed,
+                Some(config.temp),
+                Some(config.top_p),
+            ),
+            device,
+            prompt_history: String::from("<|im_start|>"),
+            image_history: Vec::new(),
+        })
+    }
+
+    /// Run the inference of the SmolVLM model with previous context added.
+    ///
+    /// # Arguments
+    ///
+    /// * `image` - The rgb8    image to generate a caption for with shape [H, W, 3]
+    /// * `prompt` - The prompt to generate a caption for
+    /// * `sample_len` - The length of the generated caption
+    /// * `stdout_debug` - Whether to print the debug information to the stdout
+    ///
+    /// # Returns
+    ///
+    /// * `caption` - The generated caption
+    pub fn inference(
+        &mut self,
+        image: Option<Image<u8, 3, CpuAllocator>>,
+        prompt: &str,
+        sample_len: usize, // per prompt
+        stdout_debug: bool,
+    ) -> Result<String, SmolVlmError> {
+        if stdout_debug {
+            std::io::stdout().flush()?;
+        }
+
+        let mut response = String::new(); // collection of tokens
+
+        let mut full_prompt = String::new();
+        if let Some(raw_img) = image {
+            let (img_patches, mask_patches, rows, cols) =
+                preprocess_image(raw_img, 1920, 384, &self.device);
+
+            let img_token = get_prompt_split_image(81, rows, cols);
+            full_prompt += "\nUser:<image>";
+            full_prompt = full_prompt.replace("<image>", &img_token);
+
+            self.image_history.push((img_patches, mask_patches));
+        } else {
+            full_prompt += "\nUser: ";
+        }
+
+        full_prompt += prompt;
+        full_prompt += "<end_of_utterance>\nAssistant:";
+        response.clear();
+        self.prompt_history += &full_prompt;
+
+        for i in 0..sample_len {
+            let encoding = self.tokenizer.encode(self.prompt_history.clone(), false)?;
+            let tokens = encoding.get_ids();
+            let input =
+                Tensor::from_slice(tokens, Shape::from_dims(&[tokens.len()]), &self.device)?;
+            let vision_data = if self.image_history.len() > 0 {
+                let image_token_mask = input.broadcast_eq(&self.image_token_tensor)?;
+                Some((
+                    image_token_mask,
+                    &self.image_history[0].0,
+                    &self.image_history[0].1,
+                ))
+            } else {
+                None
+            };
+            let logits = self.model.forward(&input, i, vision_data)?;
```

Comment:
How easy it is for us to split the forward to `forward_vision`, `forward_text`, and `forward_with_features`? So that we may save the image feature for the future queries. Consider a scenario that we are querying on one image multiple times.